File: src/p0039/python/solution.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ansList = list()
        ans = list()
        def dfs(i: int, remaining: int):
            if i == len(candidates):
                return
            if remaining == 0:
                ansList.append(ans.copy())
                return
            dfs(i+1, remaining)
            if remaining - candidates[i] >= 0:
                ans.append(candidates[i])
                dfs(i, remaining - candidates[i])
                ans.pop()
        dfs(0, target)
        
        return ansList

File: src/p0039/golang/solution.go
func combinationSum(candidates []int, target int) [][]int {
	var (
		dfs func(i, remaining int)
		ans [][]int
		t   []int
	)
	n := len(candidates)
	dfs = func(i, remaining int) {
		if remaining == 0 {
			ans = append(ans, append([]int(nil), t...))
			return
		}
		for j := i; j < n; j++ {
			if remaining-candidates[j] >= 0 {
				t = append(t, candidates[j])
				dfs(j, remaining-candidates[j])
				t = t[:len(t)-1]
			}
		}
	}
	dfs(0, target)

	return ans
}